pageserver/src/tenant/timeline.rs

```diff
@@ -2907,6 +2907,17 @@ impl Timeline {
                 // `create_delta_layer` will not modify the layer map.
                 // We will remove frozen layer and add delta layer in one atomic operation later.
                 let layer = self.create_delta_layer(&frozen_layer, ctx).await?;
+                // 5GB limit for objects without multipart upload (which we don't want to use)
+                // Make it a little bit below to account for differing GB units
+                // https://docs.aws.amazon.com/AmazonS3/latest/userguide/upload-objects.html
+                const S3_UPLOAD_LIMIT: u64 = 4_500_000_000;
+                if layer.metadata().file_size() > S3_UPLOAD_LIMIT {
+                    return Err(FlushLayerError::Other(anyhow::anyhow!(
```

Comment:
```suggestion
                if layer.metadata().file_size() > S3_UPLOAD_LIMIT {
                    layer.delete_on_drop();
                    return Err(FlushLayerError::Other(anyhow::anyhow!(
```

However the `struct Layer` will try to delete itself from s3, which will succeed without an issue, I wonder if there could be a related problem created here..

Alternatively:

```suggestion
                if layer.metadata().file_size() > S3_UPLOAD_LIMIT {
                    if let Err(e) = tokio::fs::remove_file(layer.local_path()).await {
                        tracing::warn!("failed to cleanup too large file: {e}");
                    }
                    return Err(FlushLayerError::Other(anyhow::anyhow!(
```

Reply:
Yeah, I didn't do the `delete_on_drop` precisely because it will cause the interaction with the upload queue
https://github.com/neondatabase/neon/blob/bf4e708646388ff1aee8eafe78c108693a53d5a9/pageserver/src/tenant/storage_layer/layer.rs#L526-L527

(Didn't we want to clean that up?)

Reply:
With this, I think we are creating more problems by moving the check upstack than with removing the tempfile on failure.